### PR TITLE
initial l5 support

### PR DESCRIPTION
**Pull request by Evidlo**
Made more progress.  I looked for anywhere `STM32_FLASH_TYPE_G4` was defined and made an equivalent section for `STM32_FLASH_TYPE_L5`.  I crossed referenced all the G0/G4 definitions in `stm32flash.h` and the associated reference manual and found equivalent values in the L5 manual.  I wasn't able to find where the value for `STM32Gx_FLASH_SR_ERROR_MASK` was taken from in the reference manual, so I just used the same value for L5.

Most of the registers/offsets in the reference manual are prefixed with "NS" (non-secure), which I've followed with my variable names.  I don't know if other STM32's distinguish NS vs PD addresses in this way, but I can remove the NS prefix if we want to keep a simple naming convention.

### CODE REVIEW

**Comment by Evidlo** [comment]
I'm having trouble getting the flash to erase. `st-flash` is detecting nonzero bytes during the write step, and I think the flash isn't being erased at all, because my blink program restarts as soon as I power-cycle the board.

```
[evan@blackbox ~] /home/evan/resources/stlink/build/Release/bin/st-flash write /tmp/arduino_build_451074/blink.ino.bin 0x8000000
st-flash 1.7.0-186-gc4762e6-dirty
Failed to parse flash type or unrecognized flash type

detected chip_id parametres

# Device Type: STM32L552
# Reference Manual: RM0438
#
chip_id 0x472
flash_type 10
flash_size_reg 0xbfa05e0
flash_pagesize 0x1000
sram_size 0x40000
bootrom_base 0xbf90000
bootrom_size 0x8000
option_base 0x0
option_size 0x0
flags 0

2022-05-17T22:09:30 INFO common.c: STM32L552: 256 KiB SRAM, 512 KiB flash in at least 4 KiB pages.
file /tmp/arduino_build_451074/blink.ino.bin md5 checksum: 11eea395bf81feee318bcef406d42c, stlink checksum: 0x001e2964
2022-05-17T22:09:30 INFO common_flash.c: Attempting to write 20568 (0x5058) bytes to stm32 address: 134217728 (0x8000000)
-> Flash page at 0x8000000 erased (size: 0x1000)
-> Flash page at 0x8001000 erased (size: 0x1000)
-> Flash page at 0x8002000 erased (size: 0x1000)
-> Flash page at 0x8003000 erased (size: 0x1000)
-> Flash page at 0x8004000 erased (size: 0x1000)
-> Flash page at 0x8005000 erased (size: 0x1000)

2022-05-17T22:09:30 INFO flashloader.c: Starting Flash write for WB/L5/G0/G4
  4/  5 pages written
2022-05-17T22:09:34 ERROR common_flash.c: Flash memory contains a non-erased value
2022-05-17T22:09:34 ERROR common_flash.c: Flash programming error: 0x00000080
stlink_fwrite_flash() == -1
```

**Review by Ant-ON — CHANGES_REQUESTED** [review]

**Comment by Ant-ON** [comment]
You can also add page size reduction if DBANK is used. Example (`if (sl->chip_flags & CHIP_F_HAS_DUAL_BANK && sl->flash_type == STM32_FLASH_TYPE_L5_U5) ...`): https://github.com/stlink-org/stlink/blob/951859c119c570f3bdf8b42da0fbbaffec0848ab/src/common.c#L280-L287

And small fix page erase code

```c
    } else if (sl->flash_type == STM32_FLASH_TYPE_L5_U5) {
      uint32_t flash_page;
      stlink_read_debug32(sl, STM32L5_FLASH_NSCR, &val);
      if (sl->flash_pgsz == 0x800 && offset >= sl->flash_size/2) {
        flash_page = (flashaddr - STM32_FLASH_BASE - sl->flash_size/2) / 
                                 (uint32_t)(sl->flash_pgsz));
        // set bank 2
        val |= (1 << STM32L5_FLASH_NSCR_NSBKER);
      } else {
        flash_page = (flashaddr - STM32_FLASH_BASE) / 
                                 (uint32_t)(sl->flash_pgsz));
        // clear bank selection / set bank 1
        val &= ~(1 << STM32L5_FLASH_NSCR_NSBKER);
      }

      // sec 6.9.9
      val &= ~(0x7F << 3);
      val |= ((flash_page & 0x7F) << 3) | (1 << STM32L5_FLASH_NSCR_NSPER);
      stlink_write_debug32(sl, STM32L5_FLASH_NSCR, val);
    }
```

File: inc/stm32flash.h [hunk 1]
```diff
@@ -207,6 +208,46 @@
...
+#define STM32L5_FLASH_NSCR (STM32L5_FLASH_REGS_ADDR + 0x28)
+#define STM32L5_FLASH_ECCR (STM32L5_FLASH_REGS_ADDR + 0x30)
+#define STM32L5_FLASH_OPTR (STM32L5_FLASH_REGS_ADDR + 0x40)
+
+// FLASH_NSCR
+// RM0438, pg. 242
+#define STM32L5_FLASH_NSCR_NSPG 0        /* Program */
+#define STM32L5_FLASH_NSCR_NSPER 1       /* Page erase */
+#define STM32L5_FLASH_NSCR_NSMER1 2      /* Bank 1 erase */
+#define STM32L5_FLASH_NSCR_NSPNB 3       /* Page number (8 bits) */
```
Comment (Ant-ON) [inline inc/stm32flash.h:227]: Typo. Page number has 7 bit.

**Review by Ant-ON — CHANGES_REQUESTED** [review]

File: inc/stm32flash.h [hunk 2]
```diff
@@ -207,6 +208,46 @@
...
+#define STM32L5_FLASH_NSSR_NSEOP 0       /* End of Operation */
+#define STM32L5_FLASH_NSSR_NSOPERR 1
+#define STM32L5_FLASH_NSSR_NSPROGERR 3
+#define STM32L5_FLASH_NSSR_NSWRPERR 4
+#define STM32L5_FLASH_NSSR_NSPGAERR 5
+#define STM32L5_FLASH_NSSR_NSSIZERR 6
+#define STM32L5_FLASH_NSSR_NSPGSERR 7
+#define STM32L5_FLASH_NSSR_OPTWERR 12
+#define STM32L5_FLASH_NSSR_BSY 16        /* Busy */
+#define STM32L5_FLASH_NSSR_ERROR_MASK (0x3fa)
```
Comment (Ant-ON) [inline inc/stm32flash.h:249]: STM32L5_FLASH_NSSR_ERROR_MASK shod be 0x20FA (see sec 6.9.7)

**Comment by Ant-ON** [comment]
@Evidlo RM0438 pg. 183 have note:
```
At power-on reset or a system reset, the main regulator voltage range 2 is selected by
default. Consequently, the voltage scaling range must be programmed to range 0 or range 1
via VOS[1:0] bits in the PWR_CR1 register prior to any Flash erase and programming
operation.
```
I would add the following code to the start of the `unlock_flash_if` function
```c
#define STM32L5_PWR_CR1 0x40007000
#define STM32L5_PWR_CR1_VOS 9 /* Voltage scaling range selection */

...
  if (sl->flash_type == STM32_FLASH_TYPE_L5_U5) {
    // Set the voltage scaling range to the range 0 to perform flash operations
    // RM0438 pg. 183
   uint32_t mask = (0x3 << STM32L5_PWR_CR1_VOS);
    stlink_read_debug32(sl, STM32L5_PWR_CR1, &val);
    if (val & mask) {
      val &= ~mask;
      stlink_write_debug32(sl, STM32L5_PWR_CR1, val);
    }
  }
```

**Comment by Evidlo** [comment]
@Ant-ON 

Why bother reading the value of the `VOS` bits?  Why not just zero them out every time?

**Comment by Ant-ON** [comment]
@Evidlo You are right. It looks somehow without strong logic. It is necessary to write the condition differently so the idea becomes clear:
```c
...
if ((val & mask) > 1) {
...
```
I wanted to make as few changes to the system as possible. `st-flash` has the ability to connect without rebooting (`--hot-plug` option)

ps Have you checked flash the firmware? Did it work?

**Comment by Nightwalker-87** [comment]
@Evidlo: Please correct the detected errors. Can we also draw this to a close?

**Comment by Evidlo** [comment]
I've made the changes, but erasing still doesn't work at there are nonzero values in the flash:

```
[evan@blackbox src] /home/evan/resources/stlink/build/Release/bin/st-flash write /tmp/arduino_build_678130/blink.ino.bin 0x8000000
st-flash 1.7.0-193-ga23ef65-dirty
Failed to parse flash type or unrecognized flash type UNKNOWN from file /usr/local/stlink/chips/unknown_device.chip

detected chip_id parametres

# Device Type: STM32L552
# Reference Manual: RM0438
#
chip_id 0x472
flash_type 10
flash_size_reg 0xbfa05e0
flash_pagesize 0x1000
sram_size 0x40000
bootrom_base 0xbf90000
bootrom_size 0x8000
option_base 0x0
option_size 0x0
flags 0

2022-06-04T21:39:05 INFO common.c: STM32L552: 256 KiB SRAM, 512 KiB flash in at least 4 KiB pages.
file /tmp/arduino_build_678130/blink.ino.bin md5 checksum: 1888c7fea96a4b962dc1fb93fda4aaf, stlink checksum: 0x001e34a6
2022-06-04T21:39:05 INFO common_flash.c: Attempting to write 20584 (0x5068) bytes to stm32 address: 134217728 (0x8000000)
-> Flash page at 0x8000000 erased (size: 0x1000)
-> Flash page at 0x8001000 erased (size: 0x1000)
-> Flash page at 0x8002000 erased (size: 0x1000)
-> Flash page at 0x8003000 erased (size: 0x1000)
-> Flash page at 0x8004000 erased (size: 0x1000)
-> Flash page at 0x8005000 erased (size: 0x1000)

2022-06-04T21:39:05 INFO flashloader.c: Starting Flash write for WB/L5/G0/G4
  4/  5 pages written
2022-06-04T21:39:09 ERROR common_flash.c: Flash memory contains a non-erased value
2022-06-04T21:39:09 ERROR common_flash.c: Flash programming error: 0x00000080
stlink_fwrite_flash() == -1
```

**Comment by Nightwalker-87** [comment]
@Evidlo: Much better now, but some issue on macos remains as can be seen above.
It's only a _-Werror,-Wshorten-64-to-32_  so it should not be too complicated to resolve this as well, I think.

File: inc/stm32flash.h [hunk 3]
```diff
@@ -207,6 +208,46 @@
 
 #define STM32L4_FLASH_OPTR_DUALBANK 21
 
+// Flash registers common to STM32L5 series
+// RM0438, pg. 241
+#define STM32L5_FLASH_REGS_ADDR ((uint32_t)0x50022000)
```
Comment (Ant-ON) [inline inc/stm32flash.h:213]: We should use the 0x40022000 base address. The 0x50022000 address used in the security mode (when the TrustZone is enabled). RM0438, pg. 2194

**Review by Ant-ON — CHANGES_REQUESTED** [review]

**Comment by Ant-ON** [comment]
@Evidlo I seem to have found the problem. Can you fix and try flash?

**Comment by Evidlo** [comment]
@Ant-ON, I've made the changes, but I'm getting an identical error message.

Also an unrelated problem is that the `--reset` argument doesn't seem to work, but running the `reset` command separately before writing does work.

<details>
<summary>--reset argument</summary>

```
[evan@blackbox ~] ~/resources/stlink/build/Release/bin/st-flash --reset write ~/resources/eduplsb/fw/blink.ino.bin 0x8000000
st-flash 1.7.0-197-g3d67d23
Failed to parse flash type or unrecognized flash type UNKNOWN from file /usr/local/stlink/chips/unknown_device.chip

detected chip_id parametres

# Device Type: STM32L552
# Reference Manual: RM0438
#
chip_id 0x472
flash_type 10
flash_size_reg 0xbfa05e0
flash_pagesize 0x1000
sram_size 0x40000
bootrom_base 0xbf90000
bootrom_size 0x8000
option_base 0x0
option_size 0x0
flags 0

2022-06-15T16:26:35 INFO common.c: STM32L552: 256 KiB SRAM, 512 KiB flash in at least 4 KiB pages.
file /home/evan/resources/eduplsb/fw/blink.ino.bin md5 checksum: 1888c7fea96a4b962dc1fb93fda4aaf, stlink checksum: 0x001e34a6
2022-06-15T16:26:35 INFO common_flash.c: Attempting to write 20584 (0x5068) bytes to stm32 address: 134217728 (0x8000000)
2022-06-15T16:26:35 ERROR common_flash.c: Flash programming error: 0x00000080
2022-06-15T16:26:35 WARN common_flash.c: Failed to erase_flash_page(0x8000000) == -1
2022-06-15T16:26:35 ERROR common_flash.c: Failed to erase the flash prior to writing
stlink_fwrite_flash() == -1
```

</details>

<details>
<summary>reset command</summary>

```
[evan@blackbox ~] ~/resources/stlink/build/Release/bin/st-flash reset
st-flash 1.7.0-197-g3d67d23
Failed to parse flash type or unrecognized flash type UNKNOWN from file /usr/local/stlink/chips/unknown_device.chip

detected chip_id parametres

# Device Type: STM32L552
# Reference Manual: RM0438
#
chip_id 0x472
flash_type 10
flash_size_reg 0xbfa05e0
flash_pagesize 0x1000
sram_size 0x40000
bootrom_base 0xbf90000
bootrom_size 0x8000
option_base 0x0
option_size 0x0
flags 0

2022-06-15T16:24:29 INFO common.c: STM32L552: 256 KiB SRAM, 512 KiB flash in at least 4 KiB pages.
2022-06-15T16:24:29 WARN common.c: NRST is not connected
[evan@blackbox ~] ~/resources/stlink/build/Release/bin/st-flash write ~/resources/eduplsb/fw/blink.ino.bin 0x8000000
st-flash 1.7.0-197-g3d67d23
Failed to parse flash type or unrecognized flash type UNKNOWN from file /usr/local/stlink/chips/unknown_device.chip

detected chip_id parametres

# Device Type: STM32L552
# Reference Manual: RM0438
#
chip_id 0x472
flash_type 10
flash_size_reg 0xbfa05e0
flash_pagesize 0x1000
sram_size 0x40000
bootrom_base 0xbf90000
bootrom_size 0x8000
option_base 0x0
option_size 0x0
flags 0

2022-06-15T16:24:33 INFO common.c: STM32L552: 256 KiB SRAM, 512 KiB flash in at least 4 KiB pages.
file /home/evan/resources/eduplsb/fw/blink.ino.bin md5 checksum: 1888c7fea96a4b962dc1fb93fda4aaf, stlink checksum: 0x001e34a6
2022-06-15T16:24:33 INFO common_flash.c: Attempting to write 20584 (0x5068) bytes to stm32 address: 134217728 (0x8000000)
-> Flash page at 0x8000000 erased (size: 0x1000)
-> Flash page at 0x8001000 erased (size: 0x1000)
-> Flash page at 0x8002000 erased (size: 0x1000)
-> Flash page at 0x8003000 erased (size: 0x1000)
-> Flash page at 0x8004000 erased (size: 0x1000)
-> Flash page at 0x8005000 erased (size: 0x1000)

2022-06-15T16:24:34 INFO flashloader.c: Starting Flash write for WB/L5/G0/G4
  5/  5 pages written
2022-06-15T16:24:38 ERROR common_flash.c: Flash memory contains a non-erased value
2022-06-15T16:24:38 ERROR common_flash.c: Flash programming error: 0x00000080
stlink_fwrite_flash() == -1
```

</details>

**Comment by Evidlo** [comment]
@Ant-ON If you'd like it, I can also give you guest access to a machine that has the board connected if you give me your ssh pubkey.

**Comment by Nightwalker-87** [comment]
@Evidlo What is the current state here? Is there any reasonable perspective to draw this topic to a close?

**Comment by Nightwalker-87** [comment]
Note: This unfinished PR will be closed in 7 days, if it remains incomplete.

**Comment by Evidlo** [comment]
@Ant-ON Any additional ideas for how to debug this?

**Comment by Ant-ON** [comment]
@Evidlo For starters, I think it's worth checking the memory erasing process. To do this, we will erase the memory and read its contents:
```bush
st-flash erase 0x8000000 0x6000
st-flash read ./temp.bin 0x8000000 0x6000
```
If the erasing process works correctly, then the `temp.bin` file will have filled by `0xFF` bytes

**Comment by Ant-ON** [comment]
@Evidlo To fix the reset after the flashing, you need to change the reset code in the src/st-flash/flash.c file to
```c
    if (o.reset) {
        stlink_reset(sl, RESET_AUTO);
        stlink_run(sl, RUN_NORMAL);
    }
```

File: config/chips/L5x2.chip [hunk 4]
```diff
@@ -0,0 +1,14 @@
+# Chip-ID file for STM32L5x2 device
+#
+dev_type STM32L5x2
+ref_manual_id 0438
+chip_id 0x472                // RM0438, pg. 2157
+flash_type L5_U5             // ???
```
Comment (Nightwalker-87) [inline config/chips/L5x2.chip:6]: This info is necessary.

Comment (Nightwalker-87) [inline config/chips/L5x2.chip:6]: Please complete the missing comments according to the formatting scheme (see other chip-id-files), or remove them if there is no setting such as for the option-parametres.

**Review by Nightwalker-87 — CHANGES_REQUESTED** [review]
See inline-comment. Please also finally address the mentioned topic(s) from the discussion.

**Comment by Nightwalker-87** [comment]
This attempt failed to come to an end within a reasonable timespan.
Accordingly this PR is finally closed.

**Pull request closed without merging**